chain/network/src/network_protocol/tests.rs

```diff
@@ -170,3 +172,116 @@ fn serialize_deserialize() -> anyhow::Result<()> {
 
     Ok(())
 }
+
+fn make_block_approval_message() -> RoutedMessage {
+    let mut rng = make_rng(19385389);
+    let signer = data::make_validator_signer(&mut rng);
+    let approval = Approval::new(CryptoHash::default(), 1, 1, &signer);
+    data::make_routed_message(
+        &mut rng,
+        TieredMessageBody::T1(Box::new(T1MessageBody::BlockApproval(approval))),
+    )
+}
+
+fn make_chunk_request_message() -> RoutedMessage {
+    let mut rng = make_rng(19385389);
+    data::make_routed_message(
+        &mut rng,
+        T2MessageBody::PartialEncodedChunkRequest(PartialEncodedChunkRequestMsg {
+            chunk_hash: CryptoHash::default().into(),
+            part_ords: vec![],
+            tracking_shards: Default::default(),
+        })
+        .into(),
+    )
+}
+
+#[test]
+fn test_message_matching_hashes() {
+    let message_v3 = make_chunk_request_message();
+    let message_v1: RoutedMessage = message_v3.clone().msg_v1().into();
+    assert_eq!(message_v3.hash(), message_v1.hash());
+}
+
+#[test]
+fn test_upgrading_to_v3() {
+    let message_v3 = make_chunk_request_message();
+    let message_v1: RoutedMessage = message_v3.msg_v1().into();
+    let message_v2: RoutedMessage =
+        RoutedMessageV2 { msg: message_v1.clone().msg_v1(), created_at: None, num_hops: 0 }.into();
+
+    // Tier 2
+    let mut message = message_v1.clone();
+    assert!(matches!(message, RoutedMessage::V1(_)));
+    let _ = message.num_hops_mut();
+    assert!(matches!(message, RoutedMessage::V3(_)));
+
+    assert!(matches!(message_v1, RoutedMessage::V1(_)));
+    let body = message_v1.body_owned();
+    assert!(matches!(body, TieredMessageBody::T2(_)));
+
+    let mut message: RoutedMessage = message_v2.clone();
+    assert!(matches!(message, RoutedMessage::V2(_)));
+    let _ = message.num_hops_mut();
+    assert!(matches!(message, RoutedMessage::V3(_)));
```

Comment:
Please add some comments explaining what we are testing. Specifically this line (do we convert everything to V3 on mutation?)